src/FFMpeg/Media/ComplexMedia.php

```diff
@@ -0,0 +1,424 @@
+            }
+            if ($format->getAudioKiloBitrate() !== null) {
+                $commands[] = '-b:a';
+                $commands[] = $format->getAudioKiloBitrate() . 'k';
+            }
+            if ($format->getAudioChannels() !== null) {
+                $commands[] = '-ac';
+                $commands[] = $format->getAudioChannels();
+            }
+        }
+
+        // If the user passed some extra parameters.
+        if ($format->getExtraParams()) {
+            $commands = array_merge($commands, $format->getExtraParams());
+        }
+
+        return $commands;
+    }
+
+    /**
+     * @param ComplexFilterInterface $filter
+     *
+     * @return string
+     */
+    private function applyComplexFilter(ComplexFilterInterface $filter)
+    {
+        /** @var $format VideoInterface */
+        $filterCommands = $filter->applyComplex($this);
+        foreach ($filterCommands as $index => $command) {
+            if ($command === '-vf' || $command === '-filter:v' || $command === '-filter_complex') {
+                unset ($filterCommands[$index]);
+            }
+        }
+
+        $strCommand = implode(' ', $filterCommands);
+
+        // Compatibility with the some existed filters:
+        // If the command contains [in], just replace it to inLabel. If not - to add it manually.
+        if (stripos($strCommand, '[in]') !== false) {
+            $strCommand = str_replace('[in]', $filter->getInLabels(), $strCommand);
+            $in = '';
+        } else {
+            $in = $filter->getInLabels();
+        }
+
+        // If the command contains [out], just replace it to outLabel. If not - to add it manually.
+        if (stripos($strCommand, '[out]') !== false) {
+            $strCommand = str_replace('[out]', $filter->getOutLabels(), $strCommand);
+            $out = '';
+        } else {
+            $out = $filter->getOutLabels();
+        }
+
+        return $in . $strCommand . $out;
+    }
+
+    /**
+     * @return void
+     * @throws RuntimeException
+     */
+    protected function assertFiltersAreCompatibleToCurrentFFMpegVersion()
+    {
+        $messages = array();
+        $currentVersion = $this->getFFMpegDriver()->getVersion();
+        /** @var ComplexFilterInterface $filter */
+        foreach ($this->filters as $filter) {
+            if (version_compare($currentVersion, $filter->getMinimalFFMpegVersion(), '<')) {
+                $messages[] = $filter->getName() . ' filter is supported starting from '
+                    . $filter->getMinimalFFMpegVersion() . ' ffmpeg version';
+            }
+        }
+
+        if (!empty($messages)) {
+            throw new RuntimeException(implode('; ', $messages)
+                . '; your ffmpeg version is ' . $currentVersion);
+        }
+    }
+
+    /**
+     * @return array
+     */
+    protected function buildCommand()
+    {
+        $globalOptions = array('threads', 'filter_threads', 'filter_complex_threads');
```

Comment:
Could you please document the newly created `filter_threads` and `filter_complex_threads` as well? :)

Reply:
Hmm, those options are documented in the official man of the ffmpeg: https://ffmpeg.org/ffmpeg.html#Advanced-options

I just added the possibility of the configuration them for user.

Do you want me to add this information into the readme?

Reply:
@CaliforniaMountainSnake would then not be `AdvancedMedia` be a better name when they are documented as advanced option in ffmpeg? /cc @jens1o

Reply:
I agree that the naming scheme should be as consistent as possible.